### DISCUSSION TITLE
Adding calibration information to SwzDeviceFeed

### DISCUSSION THREAD

**Discussion by benafischer94** [usdot-jpo-ode/jpo-wzdx#262] [Ideas] [answered]
# Proposal

Given that Smart Work Zone Devices, or more generically, ITS devices include some that require
calibration for valid data, e.g.: detection, cameras, various sensors; adding fields for calibration
would make sense.  I would propose something along the lines of the following as an optional field
collection for devices that regularly require calibration.

## Initial Proposal for Format

Format is not in full schema formatting, but should provide a basis to kick off the conversation
for further debate around what the fields, and potential acceptable values are. I would caution
against being all encompassing initially, and instead allow a "good enough" approach to be acceptable
as the format evolves and is further used it should naturally lead to full coverage.

```json
{
  "calibration": {
    "last_calibrated" : "datetime+tz",
    "manual_calibration": "boolean",
    "calibrated_by": "Organization that is calibrating rather than individual within an organization",
    "calibration_standard": "url",
    "next_calibration_recommended": "datetime+tz",
    "device_malfunctions": {
      "malfunctions": [
        {
          "malfunction_type": [
            {
              "description": "a grouping of acceptable malfunction types should accompany e.g. 'pan-tilt':'boolean', 'detection':'boolean', 'led:'boolean'",
              "severity": "integer",
            }
          ],
        },
      ],
      "functionality_impacted":[
        {
          "name": "generalize name of functionality",
          "description": "one of an acceptable set of functionalities based on device. e.g: led_out, led_stuck_on, coupled with what is affected"
        }
      ],
      "total_severity": "integer"
    }
  }
}
```

## Rationale

This will allow organizations to determine the reliability of data coming into system. e.g. an auto
manufacturer may require a higher level of confidence in detection devices than a parking lot
operator. Additionally, help assist in keeping operators honest about malfunctions in devices as well
as initial deployment being verified.

A growing number of OEMs are expressing concern of devices being improperly deployed to the field
reducing the actual impact. This leads to safety theater, organizations implementing ITS devices
with no accountability of proper deployment to effectively check a box with contract owners. It
should hopefully help provide a structure for allowing adoption of requirements on calibration of
devices to be properly certified. This could be a multi enforcement effort from OEMs, FHWA, USDOT,
State DOTs, down to municipalities and private operators.

This makes the inclusion of the calibration_standard an important part so that it is clear to installers
and operators what the expected calibration standard is. This should also lead to transparency within
contracts as to what the standards are that are being adhered to for third party validation.

Additionally, this can help lead to further creation of automatic device health-checking capabilities.
e.g: a higher confidence device can be used to provide baseline in corridor detection services.

## Field clarification

* Last Calibrated: a timestamp of when the device was calibrated. For manual calibrations where technicians
may not be entering the data immediately on the side of the road, the date with zeroed out hours should
be acceptable. For machine based calibration, an exact timestamp makes sense.

* Manual Calibration: A true false to determine if human calibrated or done by machine

* Calibrated by: The organization doing the calibration and certification. It would be reasonable for
organizations to track the individual doing it, but not to publish the exact individual for all.

* Calibration standard: Manufacturer and organizations calibration standards vary, by publishing the
exact standard followed, ambiguity can be eliminated. OEMs should make a linkable version for their
devices that they recommend public for linking and best practice adhesion.

* Device malfunctions: Some devices may have multiple capability e.g. a PDMS with detection. The detection
is a secondary function, and could be malfunctioning while the PDMS is still considered good. so an array
to support all possible malfunctions

  * Malfunction type: A generic set of lists that can be determined across a device of a given type. E.g. LEDs
for PDMS, speed variance for detection.

  * Severity: A number that can be assigned to the impact of the individual malfunction. Working group
could come up with standard based severity for malfunctions of primary, secondary, and tertiary function
of components on a device of a given type.

* Functionality impacted: Some standardized items impact of the specific malfunction to devices. This
can be regardless of primary purpose of the device. An array of impacted functions

  * Name: Generic name of functionality e.g. message viewability, detection

  * Description: Generic descriptions to accompany this for reporting what it means to end-users.

* Total severity: A total severity for all individual malfunctions. Organizations can make decisions
and set thresholds for usability of data based on the total severity number.

## Call To Action

Initial implementation would be optional for all devices to allow a period of adoption, either based
on time or major version revision. This should reduce heartburn associated with making a requirement
for calibration of devices.

Given the high priority and push for devices to be deployed into work zones, making sure that they are
reporting good actionable data should be accompanying this push. Along with the rapid development that
the WZDX has demonstrated, a draft proposal should be possible within a three to six month period.

## Justification / Resources

A few examples of systems not working due to failures in comms, calibration, etc. As well as existing
data feeds use verification confidence. The quiet part not always acknowledged out-loud is that the majority
of devices work well when calibrated, but if no one is watching, it gets skipped occasionally. Redeployment
due to changing work zones can exasperate this problem. These sources aren't exhaustive, and there admittedly
is a lot of anecdotal evidence as well. Though there is no shortage of documentation to support improperly
calibrated/deployed devices if individuals request additional it can be presented.

* [England smart system failure](https://www.tech-gate.org/usa/2021/08/29/smart-motorway-system-is-nicknamed-die-now-by-horrified-staff-following-computer-crashes/)
* [TPIMS Data Feed specification](https://trucksparkhere.com/wp-content/uploads/2019/01/TPIMS_TruckParking_Data_Interface_App_Developers_V1.1.pdf)
* [Data sources used in RCA](https://fdotwww.blob.core.windows.net/sitefinity/docs/default-source/research/reports/fdot-bd544-26-rpt.pdf?sfvrsn=c1bf8089_2)

**Comment by j-d-b** [comment, marked as answer]
Thank you for the excellent organization and thorough proposal. Given the specificity of the proposal, this is better addressed in an Issue, so I am going to transfer it to an issue.

**Reply by j-d-b** [reply]
See #263.

**Reply by benafischer94** [reply]
Thanks! Makes sense, will move conversation there.

**Comment by rdsheckler** [comment]
This is parallel to something we are working on as we build our system.  We have a need to provide some measure of device integrity both to the data consumer as well as the device owner.  Because the device may appear to be working properly in the field while the connectivity piece may have problems in practice it is very hard for the owner of hundreds of devices to know where issues have arisen.  As a result we are preparing to push back to the device owner a daily list of devices that we believe that they own and the degree to which we perceive them to be in good repair.  If we report to the device owner that an arrow board seems to have a wire broken or a light not working it becomes the responsibility of the owner to fix the device and until it is fixed we plan to report it as unverified.

So, this goes to radar being improperly aimed, message boards not responding, locations being obviously wrong (middle of a lake), batteries being depleted, or the device just stopped talking.  In practice as the connectivity provider we can spot some of these things but we can't fix it.  All we can do is mark it as 'verified' or not 'verified' or in Ben's terms 'calibrated or not.
